traffic_ops/client/role.go

```diff
@@ -107,8 +107,28 @@ func (to *Session) GetRoleByID(id int) ([]tc.Role, ReqInf, int, error) {
 
 // GET a Role by the Role name
```

Comment:
this will break GoDocs. Need to look like this:

// GetRoleByName gets a a role by the role name.

Reply:
I don't think that's true. Plus, this is the pattern set by all other `func`s in that file.

Reply:
maybe i'm wrong i was just going by something @rob05c told me.

Reply:
@mitchell852 is correct. Symbol (function, variable, package, etc) comments should be a full sentence, starting with the name of the symbol. 

>this is the pattern set by all other funcs in that file.

The comments on the other funcs in the file are broken.

See https://blog.golang.org/godoc-documenting-go-code

Reply:
I don't see anything in there about "starting with the name of the symbol". All it says is

> _"The convention is simple: to document a type, variable, constant, function, or even a package, write a regular comment directly preceding its declaration, with no intervening blank line."_

And that the first sentence appears on the package list. You can see in [the generated documentation](https://godoc.org/github.com/apache/trafficcontrol/traffic_ops/client#Session.GetRoleByID) that those comments don't break anything. They're not great, but I think "broken" is a stretch.

Reply:
>Notice this comment is a complete sentence that begins with the name of the element it describes. This important convention allows us to generate documentation in a variety of formats, from plain text to HTML to UNIX man pages, and makes it read better when tools truncate it for brevity, such as when they extract the first line or sentence. 

They don't break code, no. But they're broken English, and they're broken according to the language-recommended format. GoDoc should be complete sentences, starting with the symbol name. All standard library documentation follows this format, as do all projects with well-formed GoDoc (e.g. https://godoc.org/github.com/miekg/dns).

Would you not object if our readthedocs docs were filled with sentence fragments and missing periods?

Reply:
No, it should be a complete sentence, but I didn't see anything about starting with the entity it describes (wasn't looking for 'element').

We don't support third-party GoDoc parsers, and it's clear that what he's doing works - even if not well. It's certainly much better to use a complete sentence, but I think to call something that works "broken" is a bit harsh.

Established practice for things like this - afaik from examples like "use `let`/`const` instead of `var`" or "label-able elements should have labels" - is to file an issue to fix it globally rather than worry about breaking the established pattern. Even if it is absolutely better within the scope of this code to do so.

Reply:
It's not only third party parsers. There are many first-party parsers, not just HTML. Including the command-line `go doc`, which it categorically breaks.

>Established practice for things like this - afaik from examples like "use let/const instead of var" or "label-able elements should have labels" - is to file an issue to fix it globally rather than worry about breaking the established pattern.

This isn't just "not the ideal best practice," it categorically breaks GoDoc parsing. We should not merge code into the repository that breaks GoDoc. "Following the surrounding convention" should be followed when it's a "best practice" issue, not when something is categorically broken. This is broken. It breaks GoDoc parsers, including first-party language parsers. The fact that we already have broken documentation does not make it acceptable to add more broken documentation. Certainly not considering it's trivial to fix.

>I think to call something that works "broken" is a bit harsh.

It is broken. It does not work. 

Notwithstanding, it would be broken even if `go doc` didn't break. Just because something coincidentally produces the right output does not mean it isn't "broken". If I write a function to compute powers, which actually squares, but I only ever call it with 2, the function is still broken.

https://golang.org/doc/effective_go.html#commentary

```
 Doc comments work best as complete sentences, which allow a wide variety of automated presentations. The first sentence should be a one-sentence summary that starts with the name being declared.

// Compile parses a regular expression and returns, if successful,
// a Regexp that can be used to match against text.
func Compile(str string) (*Regexp, error) {

If every doc comment begins with the name of the item it describes, you can use the doc subcommand of the go tool and run the output through grep. Imagine you couldn't remember the name "Compile" but were looking for the parsing function for regular expressions, so you ran the command,

$ go doc -all regexp | grep -i parse

If all the doc comments in the package began, "This function...", grep wouldn't help you remember the name. But because the package starts each doc comment with the name, you'd see something like this, which recalls the word you're looking for.

$ go doc -all regexp | grep -i parse
    Compile parses a regular expression and returns, if successful, a Regexp
    MustCompile is like Compile but panics if the expression cannot be parsed.
    parsed. It simplifies safe initialization of global variables holding
$
```

Running their example on our code:

```
2019-08-06 15:31:11 pc ~/lang/go/src/github.com/apache/trafficcontrol/traffic_ops/client
$ cat role.go | grep -B 1 DeleteRoleByID
// DELETE a Role by ID
func (to *Session) DeleteRoleByID(id int) (tc.Alerts, ReqInf, int, error) {
2019-08-06 15:31:13 pc ~/lang/go/src/github.com/apache/trafficcontrol/traffic_ops/client
$ go doc | grep -i delete
2019-08-06 15:31:15 pc ~/lang/go/src/github.com/apache/trafficcontrol/traffic_ops/client
$ 
```

Reply:
but...
```shell
$ go doc client.DeleteRoleByID
func (to *Session) DeleteRoleByID(id int) (tc.Alerts, ReqInf, int, error)
    DELETE a Role by ID
```
Still though, the only part of GoDoc that factors into our documentation is the html parser, and this doesn't break that.

Reply:
@alexluckerman - can you please change to 

// GetRoleByName gets a a role by the role name.

Reply:
confused on this whole thread. seems pretty clear to me:

https://blog.golang.org/godoc-documenting-go-code

![image](https://user-images.githubusercontent.com/251272/62632693-24c40a00-b8f0-11e9-8d6a-9738de191b27.png)

Reply:
actually, looks like there's already one open - #3288